### PR TITLE
Add random() function

### PR DESCRIPTION
**Pull request by nils-werner**
I think a `random()` function is could be pretty handy for generating some quick testing data.

I tried several methods, like generating `.coords` and `.data` myself, and creating a `COO` object from them. In the end it turned out to be the easiest to create a linear `scipy.sparse` array and simply reshape it to the desired output shape.

### CODE REVIEW

**Comment by mrocklin** [comment]
Thanks for this.  Some notes:c

1.  This could use a test
2.  Any thoughts on using this to replace our current `random_x` function currently used in tests?  Your solution is likely far more efficient.

**Comment by nils-werner** [comment]
> 1. This could use a test

Definitely! What would you like to be tested?

> 2. Any thoughts on using this to replace our current random_x function currently used in tests?

Might be worth investigating. I didn't understand what `random_x` was doing at first glance, so I just ignored it for now.

**Comment by mrocklin** [comment]
Well, we might ensure a variety of things we know about sprase matrices.  Here are a few that come to mind:

1.  The shape matches the intended shape
2.  The dtype matches the intended dtype (for a few dtypes)
3.  The density matches the intended density (according to nnz)
4.  Two calls to the same function don't produce the same results
5.  Two calls with the same args and with the same random state seed (we'll need to pass this through) produce equivalent results

**Comment by mrocklin** [comment]
> Might be worth investigating. I didn't understand what random_x was doing at first glance, so I just ignored it for now.

It attempts to do the same thing that you're doing here, just much less efficiently, and much less generally :)

**Comment by hameerabbasi** [comment]
The proper way to test a random function is currently a hot debate topic. There seem to be two ways to test it, both which I'm not happy with:

- Mock/Monkeypatch the underlying random function to see it is getting the correct arguments. In this case, `scipy.sparse.rand`.
- Run it lots of times until it is statistically certain to be in a certain range, then test for that range.

I will also add that `random_x` returns a dense Numpy array, and this returns a sparse one.

**Comment by mrocklin** [comment]
I don't think we need to test the random number generator within this.  We can rely on SciPy/NumPy to produce a decent distribution of numbers.  I'm mostly concerned that we test basic user expectations.

File: sparse/core.py [hunk 1]
```diff
@@ -1380,3 +1380,24 @@ def _match_arrays(a, b):
...
+        Density of the generated array.
+    dtype : dtype
+        Type of the returned array values.
+
+
+    """
+    elements = np.prod(shape)
+
+    return COO.from_scipy_sparse(
+        scipy.sparse.rand(elements, 1, density, dtype=dtype)
```
Comment (hameerabbasi) [inline sparse/core.py]: Here, it would be useful to specify `format='coo'` explicitly. Might speed up the conversion.

**Comment by hameerabbasi** [comment]
> (we'll need to pass this through)
It looks like the version of [`scipy.sparse.rand`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.rand.html) is called [`scipy.sparse.random`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.random.html#scipy.sparse.random) and only supports floating point dtypes. So we might lose something if we implement this check.

**Comment by mrocklin** [comment]
> It looks like the version of scipy.sparse.rand that supports passing through a random state is called scipy.sparse.random and only supports floating point dtypes. So we might lose something if we implement this check.

That's unfortunate

File: sparse/core.py (hunk 1 above)
Comment (nils-werner) [reply sparse/core.py]: `scipy.sparse.rand()` defaults to `format='coo'`...

**Comment by nils-werner** [comment]
> It looks like the version of scipy.sparse.rand that supports passing through a random state is called scipy.sparse.random

Really? Looking at the reference it says

    scipy.sparse.rand(m, n, density=0.01, format='coo', dtype=None, random_state=None)

and

    scipy.sparse.random(m, n, density=0.01, format='coo', dtype=None, random_state=None, data_rvs=None)

(note the `random_state` kwarg).

**Comment by hameerabbasi** [comment]
Ah, my mistake, the notes on `rand` were misleading.

> Similar function that allows a user-specified random data source.

**Comment by nils-werner** [comment]
I have replaced `random_x` and `random_x_bool` with calls to `sparse.random` in all the tests. However, I stumbled across a few oddities that need to be discussed. See code review.

File: sparse/tests/test_core.py [hunk 2]
```diff
@@ -9,28 +9,6 @@
 import sparse
```
Comment (hameerabbasi) [inline sparse/tests/test_core.py]: Please remove `random` here, it's unneeded.

File: sparse/tests/test_core.py [hunk 3]
```diff
@@ -9,28 +9,6 @@
...
-
-
-def random_x_bool(shape):
-    x = np.zeros(shape=shape, dtype=np.bool)
-    for i in range(max(5, np.prod(x.shape) // 10)):
-        x[tuple(random.randint(0, d - 1) for d in x.shape)] = True
-    return x
-
-
 @pytest.mark.parametrize('reduction,kwargs', [
```
Comment (hameerabbasi) [inline sparse/tests/test_core.py]: Add an extra newline here.

**Comment by hameerabbasi** [comment]
The issue you were having here earlier was already fixed with #56. I've added comments that should help fix the `flake8` errors.

**Comment by hameerabbasi** [comment]
I'm good to merge as soon as `flake8` is fixed.

**Comment by nils-werner** [comment]
Don't merge yet, the tests still randomly fail due to the input data

**Comment by hameerabbasi** [comment]
You might want to rebase this on master and see if the tests still fail. If we're talking about the one I saw fail (`test_reductions`) then it was fixed in #56.

**Comment by hameerabbasi** [comment]
Since it's just failing for the `np.float16` dtype, I'm pretty sure these are just floating point errors. `np.allclose` checks for a relative error of `1.0e-15`  ([see here](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html)) and `float16` is nowhere near that accurate. You might want to pass an additional argument called `rtol` to `assert_eq` that defaults to `1.0e-15`, increase that in this case. I think this fails because we're moving from straight integers stored in a floating-point `dtype` to real floating-points... Which increases the likelihood of random errors.

**Comment by nils-werner** [comment]
If I raise `atol` to `1e-2` I stop seeing random errors...

**Comment by hameerabbasi** [comment]
Great! Fix `flake8` and add a test for `random` and we should be good to merge! I could do that too, if you like!

Thanks for your work on all this.

**Comment by hameerabbasi** [comment]
`float16` has just `10` bits of floating point precision, out of which one is essentially a sign bit. That leaves `9` bits. `2 ** -9 = 1.953125e-3`. Account for the fact that the errors can accumulate, and I think we're good with `atol=1.0e-2`.

**Comment by nils-werner** [comment]
While playing around with ways of comparing arrays I realized that `COO(scipy.sparse.random)` always returns unsorted indices. Do we want them to be never sorted/always sorted/optionally sorted?

**Comment by hameerabbasi** [comment]
It would be nice to have an option `sorted` which defaults to `False`. I actually think `sorted=False` is better because we don't want our methods to assume sorted indices, we want them to work regardless of whether the input is or isn't sorted.

**Comment by hameerabbasi** [comment]
During testing I found that the `dtype` argument in `scipy.random.sparse` has no effect. I was considering removing it.

**Comment by mrocklin** [comment]
Long term we might consider copying scipy's logic to produce the coordinates but then create our own data.

**Comment by hameerabbasi** [comment]
Do you think this is good to merge? It's good from my side. I took a look at SciPy's code, it's got a lot of unnecessary branches and I think we're better off just using it.

**Comment by nils-werner** [comment]
Or we could use the coordinates from SciPy and then do

    self.data = numpy.random.rand(elements, dtype=dtype)

**Comment by hameerabbasi** [comment]
[`numpy.random.rand`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.rand.html) doesn't seem to have a dtype parameter either.

**Comment by nils-werner** [comment]
That's right... no Idea why I thought it had one...

**Comment by nils-werner** [comment]
Maybe this is starting to overengineer the solution a bit, but we could allow for a `data_callback` kwarg, and users could inject their own random data generator into the function call:

```
sparse.random((2, 3, 5), density=0.5)  # float
sparse.random((2, 3, 5), density=0.5, data_callback=lambda x: np.random.choice([True, False], size=x))  # True and False
sparse.random((2, 3, 5), density=0.5, data_callback=lambda x: np.random.randint(10, 20, size=x)).todense()  # Integers from 10 to 20
sparse.random((2, 3, 5), density=0.5, data_callback=lambda x: np.full(x, True))  # all True
```

**Comment by hameerabbasi** [comment]
No, I actually really like this solution. However, I think it should be in a separate method rather than in `sparse.random`. I think we should mimic `scipy.sparse.random` with `sparse.random`, and add another utility method `random_callback` (my brain is crapping out on a better name).

**Comment by hameerabbasi** [comment]
Merging if there are no comments until 21:00 German time. This can hold tests on other branches back, and they may need to be rebased on this, so I don't want to hold this out too long.

**Comment by nils-werner** [comment]
Actually, `scipy.sparse.random()` does exactly that, and calls the parameter `data_rvs`. However what they don't do is change the output array `dtype`.

**Comment by hameerabbasi** [comment]
Hmmm... This seems rather inefficient and wasteful to generate the data twice. Maybe it's worth it to copy [this part](https://github.com/scipy/scipy/blob/v1.0.0/scipy/sparse/construct.py#L762-L773) of the Scipy code if we can guarantee consistency internally.

**Comment by hameerabbasi** [comment]
@nils-werner Any more changes? If not, we should merge this.

**Comment by nils-werner** [comment]
Slow down... we've just doubled the LOC in `random()` a minute ago, and nobody had the time to review the code yet...

**Comment by hameerabbasi** [comment]
Good point. I just really don't want to put off changing the test code in the other branches.

File: sparse/utils.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 from .core import COO
 
 
-def assert_eq(x, y):
+def assert_eq(x, y, rtol=1.e-5, atol=1.e-8):
```
Comment (mrocklin) [inline sparse/utils.py]: Should we just pass through `**kwargs` here?

**Comment by mrocklin** [comment]
Overall this seems great to me.  It looks like we're able to handle most cases and we're staying within the scipy.sparse API.

File: sparse/utils.py (hunk 4 above)
Comment (hameerabbasi) [reply sparse/utils.py]: Yes, that seems like a better option.

**Pull request merged**